### DISCUSSION TITLE
[Web] Is there any way to include routes in web page addresses？

### DISCUSSION THREAD

**Discussion by znoliver** [AvaloniaUI/Avalonia#12346] [Q&A]
I wanted to develop a web program using Avalonia, but when I switched pages and refreshed the web page, the page went back to its original state, this is clearly wrong. I find Avalonia Web don't have router in address , therefore, the web page state cannot be stored.
Is there any way to fix this？Thanks！

**Comment by BAndysc** [comment]
You can use regular javascript API for this using javascript interop:

https://learn.microsoft.com/en-us/aspnet/core/client-side/dotnet-interop?view=aspnetcore-7.0

i.e. in main.js:
```
const dotnetRuntime = await dotnet.create();
dotnetRuntime.setModuleImports("main.js", {
    window: {
        history: {
            replaceState: (title, href) => globalThis.window.history.replaceState(null, title, href)
        },
        location: {
            href: () => globalThis.window.location.href
        }
    }
});
```

in code:
```
public partial static class Web 
{
    [JSImport("window.history.replaceState", "main.js")]
    internal static partial void ReplaceHistoryState(string title, string href);
    
    [JSImport("window.location.href", "main.js")]
    internal static partial string Href();
}
```

Now you can use Web.Href() and Web.ReplaceHistoryState() for navigation and routing.

**Comment by znoliver** [comment]
Yes, I can add router info to address ,but when I refresh the webpage , I got 404
[录屏 2023年07月27日 17时34分40秒.webm](https://github.com/AvaloniaUI/Avalonia/assets/59060212/00365445-636a-4f90-bd11-daf8fe374273)
